### ISSUE TITLE
New type inference does not support using singledispatch to decorate a function in which the first argument is a union indexed by a constrained type variable that also appears in the return type

### ISSUE THREAD

**Issue by mwchase** [python/mypy#16531]
**Bug Report**

Apologies for the wordy title, but this issue appears to be somewhat of a corner case.
In some unpublished code, I have a `singledispatch` function that doesn't pass the new type inference code.
It appears that the problem is that it's going from a **union** which depends on a **constrained typevar** to a type that **also depends on that typevar**.
If it's not a union, not constrained, or doesn't appear in the reurn type, it passes fine.

**To Reproduce**

```python
from functools import singledispatch, singledispatchmethod
from typing import AnyStr, TypeVar


T = TypeVar("T")


@singledispatch
def func(arg: list[AnyStr] | None) -> AnyStr:
    raise NotImplementedError

@singledispatch
def func_no_union(arg: list[AnyStr]) -> AnyStr:
    raise NotImplementedError

@singledispatch
def func_no_return(arg: list[AnyStr] | None) -> None:
    raise NotImplementedError

@singledispatch
def func_unconstrained(arg: list[T] | None) -> T:
    raise NotImplementedError
```
https://mypy-play.net/?mypy=master&python=3.11&gist=3bd3f48cd0e1aca5dcdd03cfd44d3d3e

(The use of `list` does not trigger the bug, but I figured it'd make things less ambiguous in the unconstrained case. Also, it can be any union. The union that triggered this issue in the first place did not include `None`. Also, `singledispatchmethod` on a similarly-typed method will also trigger the issue.)

**Actual Behavior**

Passes successfully pre-1.7.0 so long as `--new-type-inference` is not passed.

Otherwise, decorating `func` fails. All others succeed.

```
main.py:8: error: Argument 1 to "singledispatch" has incompatible type "Callable[[list[AnyStr] | None], AnyStr]"; expected "Callable[..., Never]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.7.0 and master (failed), 1.6.1 (passed with old inference)
- Python version used: 3.11
